Add Sliding Window Summation (#5946)
cpinitiative/usaco-guide: solutions/silver/usaco-1544.mdx

---
id: usaco-1544
source: USACO Silver 2026 First Contest
title: Sliding Window Summation
author: Aryaveer Shishodia
---

[Official Analysis (Python)](http://usaco.org/current/data/sol_prob3_silver_season26contest1.html)

## Explanation
The "remainder when the number of ones in the length-K window of $b$ with leftmost index $i$ is divided by two" is analogous to the XOR of the window. The only difference between the windows of $r_i$ and $r_{i+1}$ is that $r_i$'s window contains $b_i$, while $r_{i+1}$'s contains $b_{i+k}$.

$$r_i\oplus r_{i+1} = b_i\oplus b_{i+k}$$

> Since we are given $r$, choosing the value of $b_i$ determines $b_{i+k}$, which in turn determines $b_{i+2k}$, $b_{i+3k}$, and so on. This splits $b$ into $k$ binary chains, each determined by one of the first $k$ characters of $b$. For each chain, we can set the first character to `0` and `1`, and compare which one will minimize the amount of `1` bits.

Reviewer: ```suggestion
Since we are given $r$, choosing the value of $b_i$ determines $b_{i+k}$, which in turn determines $b_{i+2k}$, $b_{i+3k}$, and so on. This splits $b$ into $k$ binary chains, each dependent on one of the first $k$ characters of $b$. For each chain, we can set the first character to `0` or `1`, comparing which will minimize the amount of `1` bits.
```

> Since we are given $r$, choosing the value of $b_i$ determines $b_{i+k}$, which in turn determines $b_{i+2k}$, $b_{i+3k}$, and so on. This splits $b$ into $k$ binary chains, each determined by one of the first $k$ characters of $b$. For each chain, we can set the first character to `0` and `1`, and compare which one will minimize the amount of `1` bits.

Reviewer: Just a thought but a visualization/image for what those binary chains look like could be helpful, like say something like k = 3
R _ _ _ R _ _ _ R ...

Author: I am thinking of a diagram like this:
```
n=11, k=3
r:        1 0 0 1 1 0 0 0 1
Indices:  0 1 2 3 4 5 6 7 8 9 10
          |     |     |     |
Chain 0:  1 --> 0 --> 0 --> 0
Chain 1:    0 --> 0 --> 1 --> 0
Chain 2:      0 --> 1 --> 0
b:        1 0 0 0 0 1 0 1 0 0 0
```
We process each chain independently, counting the number of `1` bits if we start with `0`. For Chain 0, starting with `0` would produce more `1`s than starting with `1`, so we flip the first bit of Chain 0. This flips the entire chain to minimize the total number of `1` bits. Chains 1 and 2 remain unchanged since starting with `0` already gives the fewest `1`s.

I would rather get some feedback first before adding it

> The only thing that restricts the first $k$ bits of $b$ is $r_1$, which will sometimes force us to flip one of our initial bits despite it increasing the number of `1` bits in the chain. We can check if our configuration satisfies $r_1$ by recording the bitwise XOR of the optimal first $k$ bits, and storing a running minimum $m$ the minimal cost of flipping a chain. If the bitwise XOR of the first k elements doesn't align with $r_1$, we can then flip that chain back and add $m$ to our count.

Reviewer: I just wanted a second opinion on this, but the phrasing "bitwise XOR of the optimal first $k$ bits" just feels a bit awkward if anyone has any suggestions.
